Add /spawn command

MoNTE48 opened the pull request:
Adds the command /spawn and /setspawn, respectively.
This is from some WTFPL code and slightly modified to work with new MT versions. The code simply changes the static_spawnpoint setting. It must be on every server.
There are mods that do this in some other way, but it doesn't make sense - all we want from the /spawn command is to return to the spawn point. Just like from the /home command, we are waiting for a return to our home.

DS-Minetest commented on builtin/game/chatcommands.lua:
```diff
@@ -1056,3 +1056,44 @@ core.register_chatcommand("kill", {
+	func = function(name, param)
+		local player = minetest.get_player_by_name(name)
+		if not player then
+			return false, "Player not found"
```
It's not clear, which player this is referring to. I suggest something like:
```suggestion
			return false, "Command called by invalid player"
```

DS-Minetest commented on builtin/game/chatcommands.lua:
```diff
@@ -1056,3 +1056,44 @@ core.register_chatcommand("kill", {
 	end,
 })
+
+local spawn_spawnpos = minetest.setting_get_pos("static_spawnpoint")
```
```suggestion
local spawn_spawnpos = core.setting_get_pos("static_spawnpoint")
```
Not only here but also everywhere else.

DS-Minetest commented on builtin/game/chatcommands.lua:
```diff
@@ -1056,3 +1056,44 @@ core.register_chatcommand("kill", {
+		local y = math.floor(pos.y)
+		local z = math.floor(pos.z)
+		local pos_string = x..","..y..","..z
+		local pos_string_2 = "Setting spawn point to ("..x..", "..y..", "..z..")"
```
There's `core.pos_to_string`.

DS-Minetest commented on builtin/game/chatcommands.lua:
```diff
@@ -1056,3 +1056,44 @@ core.register_chatcommand("kill", {
+		local pos = player:get_pos()
+		local x = math.floor(pos.x)
+		local y = math.floor(pos.y)
+		local z = math.floor(pos.z)
```
I suggest using `vector.round`. And before you might want to add something like 0.0001 to `pos.y`. If you stand on a full node, your y pos is a full integer + 0.5. `math.floor` would just set the position into the ground.

SmallJoker commented on builtin/game/chatcommands.lua:
```diff
@@ -1056,3 +1056,44 @@ core.register_chatcommand("kill", {
+	end,
+})
+
+minetest.register_chatcommand("setspawn", {
```
I don't see much potential in this command. Server owners can just use `/set` to update the spawn position. As a result, you may not use the variable `spawn_spawnpos` to avoid the cache being outdated.

MoNTE48 replied:
Well, I can not use 'spawn_spawnpos' variable. I think this command is simple and intuitive. /home, /sethome. /give, /giveme. /spawn, /setspawn. No one likes big command with several variables, it is difficult!

benrob0329 commented:
I feel like these specialized commands should all be moved to individual games, as each game may want to implement these differently (or not at all).

MoNTE48 commented:
The /setspawn command is only available with server privilege.
I can prohibit the use of the /spawn command in singleplayer or add spawn privilege.
How else is it possible to implement this command?

paramat commented on builtin/game/chatcommands.lua:
```diff
@@ -1056,3 +1056,40 @@ core.register_chatcommand("kill", {
+core.register_chatcommand("spawn", {
+	params = "",
+	description = "Teleport to the spawn point",
+	func = function(name, param)
```
Certainly needs a priv.

benrob0329 commented:
I mean that commands like `spawn` `home` and such should be implemented in [Minetest Game](https://github.com/minetest/minetest_game) rather than in `Builtin`. Some games might not want to allow players these kind of instant teleport commands, some might not have static spawnpoints, and some might have a more flexible system for it. By moving the commands into the game-specific code, it allows each game to do things differently.

MoNTE48 commented:
Then we have to move /teleport because it does almost the same thing!

paramat commented:
'setspawn' is already possible using /set, resetting static spawnpoint isn't done often so a dedicated command isn't needed.

'spawn' allows players to teleport, so should have the 'teleport' priv, however with the teleport priv teleporting to spawnpoint is redundant.
So this would need a new 'teleport to spawnpoint' priv, but i'm not sure there's enough need or usefulness for the functionality to be added to the engine.
Do many server owners want to allow their players to teleport to spawnpoint? It seems rather cheaty for survival, it would be used for instant travel.
So it seems 'spawn' should be left to optional mods or games to implement.

:-1:

MoNTE48 commented:
Now the spawn privilege is required. Disabled by default. Take a look at the list of available privileges. I am sure that most of this command will be useful. And if not, they simply will not added this privilege for players!

appgurueu commented:
Well, Minetest is more of an engine, and not a game (like Multicraft), so I feel this PR rather belongs to Minetest Game ?

Content of an engine should be modifiable. I want to be able to override the death formspec. But I can't. The way the death formspec looks should be left up to the game.

As should be the chatcommands and in-game mechanics.

MoNTE48 commented:
@appgurueu /spawn command is already at the discretion of the game and requires the inclusion of privileges. Therefore, it does what you say.
Customizing the screen of death is what I'm working on at MultiCraft. I will share this with Minetest, too. (For example, I want to indicate the coordinates of death in this window. And also add an indication of the cause of death, it will be fun :) But this is a completely different request.

paramat commented:
Input from server owners would be good.
Is a new 'spawn' priv allowing teleport to spawn widely useful enough to add to the engine?

ClobberXD commented:
- Not all games have the concept of a central server spawn. 
  - CTF is a great example of this - players (re)spawn at their own team base.
- Teleporting to spawn might be unfair
  - CTF comes to mind again. Let's assume CTF re-defines the chat-command to teleport players to their respective team-bases. If they're caught in a tricky situation (e.g. n vs. 1 PVP) they could just teleport to base, and they're safe from danger. Well, the additional priv might help prevent abuse to a certain extent, but this is still a concern, nonetheless.

 Keeping the aforementioned points in mind, I feel this should rather be included in MTG.

MoNTE48 commented:
Rebased, simplified, added new "**spawn**" privilege.
Like the **teleport** privilege, it can be granted by the server owner or the author of subgame.

SmallJoker commented:
Looking at player opinions this feature does not seem to be wanted, regardless on how it is implemented. Rather per-game solutions are wanted.

rubenwardy commented:
as per @SmallJoker 

Thanks for the contribution, nonetheless
